enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,375 @@
+
+### Workflow Description
+
+End users will have 2 possible variations when using this functionality,
+- Create an ingress for the user workload and a managed route is created automatically
+  by the ingress-to-route controller.
+- Create a route for the user workload and this is completely managed by the end user.
+
+Both these workflows will support integrating with third party certificate management
+solutions like cert-manager with the annotation described below.
+
+**When an Ingress is used to expose user workload**
+- The end user must have generated the serving certificate as a pre requisite
+  using third-party systems like cert-manager.
+- In cert-manager's case, the Certificate CR must be created in the same namespace
+  where the Ingress is going to be created.
+- To expose a user workload, he would need to create a new Ingress with the generated secret
+  referenced in `.spec.tls.secretName`.
+- If the secret CR that is referenced exists and has a successfully generated
+  cert/key pair generated, the ingress-to-route controller copies this certificate data 
+  into the route at `.spec.tls.certificate` and `.spec.tls.key`.
+
+**When a Route is directly used to expose user workload**
+- The end user must have generated the serving certificate as a pre requisite
+  using third-party systems like cert-manager.
+- To expose a user workload, the user would create a new Route with an annotation
+  referencing the generated secret that was created in the previous step.
+- If the secret that is referenced exists and has a successfully generated
+  cert/key pair generated, this certificate data is copied into the route at 
+  `.spec.tls.certificate` and `.spec.tls.key`.
+
+*Note*: The same workflow would apply to updating an existing Route to integrate
+with cert-manager. The new certificate data will overwrite the `.spec.tls.certificate` 
+and `.spec.tls.key` in the route CR since the annotation gets a higher preference.
+
+#### Certificate Life cycle
+
+Post integrating with third-party solutions like cert-manager, the end user can also
+disable automatic certificate management. This can be done in 2 ways,
+- The end user can delete the annotation on the Route CR.
+- The end user deletes the secret that was associated with the Certificate CR.
+
+**Deleting the annotation on the Route CR**
+The end user can delete the annotation on the Route CR, this will result
+in abondoning the cert/key pair that was generated via cert-manager and resort to
+using the default certificates that were initially generated by the cluster-ingress-operator
+i.e. clearing the `.spec.tls` on Route. 
+
+**Deleting the generated secret that contains the serving cert/key pair**
+When the end user deletes the secret containing the certificate data, the `.spec.tls` on 
+Route will be cleared. This will result in the route serving default certificates. In such
+a scenario the annotation that was added by the user will be automatically deleted by the 
```

Comment:
>The deletion of the annotation seems a little odd to me. The user has expressed its intent by adding the annotation. The intent may stay the same even if the secret reference is not there any longer. Comparing this to the Ingress case: if the secret is removed would Ingress' .spec.tls.secretName field be wiped too? I don't think so.

As for the ingress case, if the `secretName` is filed and the secret is not found the reconciliation is skipped and waits  until the secret is present (so no Route is created or the route gets deleted). And if the field is not set, it pivots to using default certs. We can't do a similar thing here since the Route is created by the user and we can skip reconciling new routes, but for updates the implementation will have gaps. Hence why I decided to clear the field since we are talking about managed certificates and the controller manages `.spec.tls` when the annotation is present. 

>Also, I'm wondering how the new controller would find out that the secret is deleted and link it back to the route whose certificate needs to be cleaned. A watch on the secrets? That would need something in the secret object to link it back to the route (annotation, status etc.).

This gives a little more idea into the implementation https://github.com/openshift/enhancements/pull/1307#discussion_r1072088428.

>Overall I think the secret removal needs to be a little more detailed in the workflow chapter.
I'll add a bit more context on this.

Reply:
Updated with the above details.